vars/yamlSubstituteVariables.groovy

```diff
@@ -0,0 +1,246 @@
+import com.sap.piper.ConfigurationHelper
+import com.sap.piper.GenerateDocumentation
+import com.sap.piper.variablesubstitution.ExecutionContext
+import com.sap.piper.variablesubstitution.DebugHelper
+import groovy.transform.Field
+
+import static com.sap.piper.Prerequisites.checkScript
+
+@Field DebugHelper logger = new DebugHelper()
+@Field String STEP_NAME = getClass().getName()
+@Field Set GENERAL_CONFIG_KEYS = []
+@Field Set STEP_CONFIG_KEYS = GENERAL_CONFIG_KEYS + [
+    /**
+     * The input Yaml data as `Object`.
+     */
+    'inputYaml',
+    /**
+     * The variables Yaml data as `Object`.
+     * Can be a `List<Map<String, Object>>` or a `Map<String, Object>` and should contain
+     * variables names and values to replace variable references contained in `inputYaml`.
+     */
+    'variablesYaml',
+    /**
+     *  An `com.sap.piper.variablesubstitution.ExecutionContext` that can be used to query
+     *  whether the script actually replaced any variables.
+     */
+    'executionContext'
+]
+
+@Field Set PARAMETER_KEYS = STEP_CONFIG_KEYS
+
+/**
+ * Step that substitutes variables in a given YAML input data object. The format to reference a variable
+ * in YAML data is to use double parentheses `((` and `))`, e.g. `((variableName))`. Variables will be replaced by
+ * values that are read from another Yaml object. The script returns a deep copy of the input YAML with all occurrences
+ * of variables replaced (if they were found in the YAML of variables). The result will be written into the `commonPipelineEnvironment`
+ * by the name `yamlSubstituteVariablesResult` and will be of type `Object`.
+ * <p>
+ * The format follows <a href="https://docs.cloudfoundry.org/devguide/deploy-apps/manifest-attributes.html#variable-substitution">Cloud Foundry standards</a>.
+ * <p>
+ * Usage: `yamlSubstituteVariables inputYaml: <yamlDataObject>, variablesYaml: <yamlDataObject> [, executionContext: context]`
+ *
+ * @param arguments - the map of arguments.
+ */
+@GenerateDocumentation
+void call(Map<String, String> arguments) {
+    handlePipelineStepErrors (stepName: STEP_NAME, stepParameters: arguments) { // synchronous closure call!
+        def script = checkScript(this, arguments)  ?: this
+
+        // load default & individual configuration
+        Map config = ConfigurationHelper.newInstance(this)
+                                        .loadStepDefaults()
+                                        .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS)
+                                        .mixinStageConfig(script.commonPipelineEnvironment, arguments.stageName ?: env.STAGE_NAME, STEP_CONFIG_KEYS)
+                                        .mixin(arguments, PARAMETER_KEYS)
+                                        .use()
+
+        Object inputYaml = config?.inputYaml
+        Object variablesYaml = config?.variablesYaml
+
+        logger.setConfig(config)
+
+        if(!inputYaml) {
```

Comment:
Changed. But I kept my original error message, since I think it is better than the generic error message (which IMO lacks the name of script that caused the issue).
I know most information is in the stack trace of the Illegal Argument Exception, still I don't think it is nice.